code/datums/emotes.dm

```diff
@@ -106,25 +106,41 @@
 		return
 
 	user.log_message(msg, LOG_EMOTE)
-	var/dchatmsg = "<b>[user]</b> [msg]"
 
 	var/tmp_sound = get_sound(user)
 	if(tmp_sound && (!only_forced_audio || !intentional) && !TIMER_COOLDOWN_CHECK(user, type))
 		TIMER_COOLDOWN_START(user, type, audio_cooldown)
 		playsound(user, tmp_sound, 50, vary)
 
-	var/user_turf = get_turf(user)
-	for(var/mob/ghost in GLOB.dead_mob_list)
-		if(!ghost.client || isnewplayer(ghost))
-			continue
-		if(ghost.stat == DEAD && ghost.client && user.client && (ghost.client.prefs.chat_toggles & CHAT_GHOSTSIGHT) && !(ghost in viewers(user_turf, null)))
-			ghost.show_message("<span class='emote'>[FOLLOW_LINK(ghost, user)] [dchatmsg]</span>")
+	if(user.ckey) // Excludes the spam from the NPC mobs.
+		user.emote_for_ghost_sight("<b>[user]</b> [msg]")
 
 	if(emote_type == EMOTE_AUDIBLE)
 		user.audible_message(msg, deaf_message = "<span class='emote'>You see how <b>[user]</b> [msg]</span>", audible_message_flags = EMOTE_MESSAGE)
 	else
 		user.visible_message(msg, blind_message = "<span class='emote'>You hear how <b>[user]</b> [msg]</span>", visible_message_flags = EMOTE_MESSAGE)
 
+
+/// Sends the given emote message for all ghosts with ghost sight enabled, excluding close enough to listen normally.
+/mob/proc/emote_for_ghost_sight(message)
+	for(var/mob/ghost as anything in GLOB.dead_mob_list)
+		if(!ghost.client || isnewplayer(ghost))
+			continue
+		if(!(ghost.client.prefs.chat_toggles & CHAT_GHOSTSIGHT))
+			continue
+		var/ghost_view = ghost.client.view
+		if(ghost.z == z)
+			if(isnum(ghost_view))
+				if(get_dist(src, ghost) < ghost_view)
+					continue
+			else
+				var/list/view_range_list = splittext(ghost_view, "x")
+				if(abs(x - ghost.x) < ((text2num(view_range_list[1]) - 1) / 2))
+					if(abs(y - ghost.y) < ((text2num(view_range_list[2]) - 1) / 2))
+						continue
```

Comment:
There's a proc for this, `in_view_range()`. Might suck a bit, could be improved by using the absolute distance rather then checking if the dist is in a range. Also you may want to use the client's `/datum/view_data` datum for this if it exists, depending if you want clients to get alerts for everything in their view, or just the stuff in like, the widescreen/standard tile set.

Reply:
Oh also proc call overhead, though I'm not sure how hot this is

Reply:
That proc is not doing what it's supposed to.
Then again, the more I look at this code, the more I see issues.
Visible messages seem to only go as far as 7 tiles around.
Should they respect the view size instead, and go further to the east/west?

Reply:
I think they should respect view size, there's a chance that pattern will cause like, widescreen "meta" complaints, but if that's an issue we should call widescreen as a concept off.

The other question is do you want to go off the client's view, or off their like, base view